sdks/python/apache_beam/io/filesystem.py

```diff
@@ -0,0 +1,384 @@
+    return result
+
+  def read(self, num_bytes):
+    if not self._decompressor:
+      raise ValueError('decompressor not initialized')
+
+    self._fetch_to_internal_buffer(num_bytes)
+    return self._read_from_internal_buffer(
+        lambda: self._read_buffer.read(num_bytes))
+
+  def readline(self):
+    """Equivalent to standard file.readline(). Same return conventions apply."""
+    if not self._decompressor:
+      raise ValueError('decompressor not initialized')
+
+    io = cStringIO.StringIO()
+    while True:
+      # Ensure that the internal buffer has at least half the read_size. Going
+      # with half the _read_size (as opposed to a full _read_size) to ensure
+      # that actual fetches are more evenly spread out, as opposed to having 2
+      # consecutive reads at the beginning of a read.
+      self._fetch_to_internal_buffer(self._read_size / 2)
+      line = self._read_from_internal_buffer(
+          lambda: self._read_buffer.readline())
+      io.write(line)
+      if line.endswith('\n') or not line:
+        break  # Newline or EOF reached.
+
+    return io.getvalue()
+
+  def closed(self):
+    return not self._file or self._file.closed()
+
+  def close(self):
+    if self.readable():
+      self._read_buffer.close()
+
+    if self.writeable():
+      self._file.write(self._compressor.flush())
+
+    self._file.close()
+
+  def flush(self):
+    if self.writeable():
+      self._file.write(self._compressor.flush())
+    self._file.flush()
+
+  @property
+  def seekable(self):
+    # TODO: Add support for seeking to a file position.
+    return False
+
+  def tell(self):
+    """Returns current position in uncompressed file."""
+    return self._uncompressed_position
+
+  def __enter__(self):
+    return self
+
+  def __exit__(self, exception_type, exception_value, traceback):
+    self.close()
+
+
+class FileMetadata(object):
+  """Metadata about a file path that is the output of FileSystem.match
+  """
+  def __init__(self, path, size_in_bytes):
+    self.path = path
+    self.size_in_bytes = size_in_bytes
+
+  def __eq__(self, other):
+    """Note: This is only used in tests where we verify that mock objects match.
+    """
+    return (isinstance(other, FileMetadata) and
+            self.path == other.path and
+            self.size_in_bytes == other.size_in_bytes)
+
+  def __hash__(self):
+    return hash((self.path, self.size_in_bytes))
+
+  def __ne__(self, other):
+    return not self.__eq__(other)
+
+  def __repr__(self):
+    return 'FileMetadata(%s, %s)' % (self.path, self.size_in_bytes)
+
+
+class abstractstatic(staticmethod):
+  __slots__ = ()
+
+  def __init__(self, function):
+    super(abstractstatic, self).__init__(function)
+    function.__isabstractmethod__ = True
+
+  __isabstractmethod__ = True
+
+
+class FileSystem(object):
+  """A class that defines that can be performed on a filesystem.
+
+  All methods are abstract and they are for file system providers to
```

Comment:
Please add more detailed documentation to filesystem and filesystem_util modules. It should be fine to copy+adapt documentation from corresponding Java classes.